src/components/views/sidebar/TreeView/InstanceTreeView.js

```diff
@@ -0,0 +1,384 @@
+    padding: '0.5rem',
+  },
+}));
+
+const popperPaperProps = {
+  style: {
+    width: '16.25rem',
+    height: '17.5rem',
+    padding: '0.5rem',
+    background: popperBG,
+    boxShadow:
+      '0 0.5rem 0.5rem -0.25rem rgba(24, 24, 24, 0.03), 0 1.25rem 1.5rem -0.25rem rgba(24, 24, 24, 0.08',
+    borderRadius: '0.75rem',
+    inset: '1rem auto auto 0 !important',
+    borderWidth: '1px',
+    borderColor: dropdownBorderColor,
+    borderStyle: 'solid',
+  },
+};
+
+const initialRightClickStateCreator = () => ({
+  mouseX: null,
+  mouseY: null,
+});
+
+const InstancesTreeView = (props) => {
+  const { datasets } = props;
+  const classes = useStyles();
+
+  const [right, setRight] = useState(() => initialRightClickStateCreator());
+  const open = Boolean(right.mouseY);
+
+  const [selectedNodes, setSelectedNodes] = useState([]);
+  const [selectedNodeId, setSelectedNodeId] = useState(null);
+  const [nodes, setNodes] = useState([]);
+  const [items, setItems] = useState(datasets);
+
+  const onNodeSelect = (e, nodeId) => {
+    setSelectedNodeId(nodeId);
+  };
+
+  const onNodeToggle = (e, nodeIds) => {
+    //  TODO uncommented after proper test
+    // if (nodeIds.length === 0) {
+    //   return;
+    // }
+
+    if (nodes.length !== nodeIds.length && nodes[0] === nodeIds[0]) {
+      var original = [...nodes];
+      var newPath = [...nodeIds];
+      while (original[0] === newPath[0]) {
+        original.shift();
+        newPath.shift();
+      }
+      nodeIds = original;
+    }
+
+    setNodes(nodeIds);
+  };
+
+  function handleClick(e, nodes_ids) {
+    if (e.target.className == 'MuiTreeItem-label') setSelectedNodes(nodes_ids);
+  }
+
+  function onRightClick(event, node) {
+    event.preventDefault();
+    event.stopPropagation();
+    if (node.type === 'COMPOSITION') {
+      setRight({
+        mouseX: event.clientX - 360,
+        mouseY: event.clientY - 8,
+      });
+    }
+  }
+
+  function onClose() {
+    setRight(initialRightClickStateCreator());
+  }
+
+  // Initialize state in this hook
+  useEffect(() => {
+    // Populate tree items state with datasets
+    if (items.length === 0 && datasets.length > 0) {
+      setItems(datasets);
+    } else if (datasets.length > 0 && items.length !== datasets.length) {
+      // Update datasets, after adding a new dataset
+      setItems(datasets);
+    }
+  }, [datasets, items.length]);
+
+  const getTreeItemsFromData = (treeItems) => {
+    if (Array.isArray(treeItems) && treeItems.length <= 0) return;
+
+    return treeItems?.map((treeItemData) => {
+      let items = undefined;
+      if (treeItemData.items && treeItemData.items.length > 0) {
+        items = getTreeItemsFromData(treeItemData.items);
+      }
+
+      const hidden =
+        !nodes.includes(treeItemData?.id) &&
+        selectedNodeId !== treeItemData?.id &&
+        treeItemData?.type === 'COMPOSITION';
+
+      const labelIcon =
+        treeItemData?.type === 'MECHANISM' ? (
+          <CircleIcon />
```

Comment:
instead of this if then else you could also inject the correct react component based on a component list that contains the constants. e.g. (not tested, just an example how you could do it)
```
const MECHANISM = 'MECHANISM'
const COMPOSITION = 'COMPOSITION'
const DEFAULT_COMPONENT = 'DEFAULT COMPONENT'
const COMPONENT_ICONS = {
  MECHANISM: <CircleIcon />,
  COMPOSITION: <FileIcon />,
  DEFAULT_COMPONENT: <ShapeArrowToolIcon />
}
```
and then use it like this:
```
const labelIcon = COMPONENT_ICONS [treeItemData.type] || COMPONENT_ICONS [DEFAULT_COMPONENT]
```

Reply:
This is way better Thank you

Reply:
the constant was implemented,  but the object labelIcon didn't work as expected because where it was used is a callback function